### ISSUE TITLE
Working with HCP 'CIFTI' parcellation

### ISSUE THREAD

**Issue by griegner** [netneurolab/neuromaps#105]
### Description of issue

Thank you for developing and maintaining such a great tool!

I am working with the HCP rfMRI dataset, which includes timeseries from brain regions defined by a 300-component ICA parcellation (CIFTI `dlabel.nii` file). For this parcellation file, I want to define an instance of the `neuromaps.parcellate.Parcellator` class.

**issue**

There should be 106 surface regions after converting CIFTI file to GIFTI format. However, relabeling the GIFTI image splits in two the regions that cross left-right hemispheres, resulting in 178 unique regions. In my convoluted (!!) code example below, I try to revert this left-right split with averaging. 

Am I missing a much easier way to do this? If not, for GIFTI images, is it safe to assume that label 0 represents the background and indexing starts with 1?

**steps to reproduce issue**

[dlabel parcellation file](https://github.com/netneurolab/neuromaps/files/11620203/nregions-300_cifti.dlabel.nii.zip)

```python
import numpy as np
# neuromaps version 0.0.3+24.gabf5a5c
from neuromaps import datasets, images, parcellate


def get_lr_intersections(gifti, gifti_relabeled):
    """get the indices of regions that cross both L-R hemispheres."""
    hemi_intersection = np.intersect1d(gifti[0].agg_data(), gifti[1].agg_data())
    l_mask = np.isin(gifti[0].agg_data(), hemi_intersection)
    r_mask = np.isin(gifti[1].agg_data(), hemi_intersection)
    l_idx = np.unique(gifti_relabeled[0].agg_data()[l_mask])
    r_idx = np.unique(gifti_relabeled[1].agg_data()[r_mask])
    return l_idx[1:] - 1, r_idx[1:] - 1  # ?? ignore 0 label and reindex to start at 0 ??


# cifti ICA parcellation file
dlabel = "./nregions-300_cifti.dlabel.nii"

# cifti to surface removes volumetric data: 300 -> 106 regions
gifti = images.dlabel_to_gifti(dlabel)

# make label IDs consecutive across hemispheres: 106 -> 178 regions
gifti_relabeled = images.relabel_gifti(gifti)

# get intersections of gifti_relabeled: 72 regions
l_idx, r_idx = get_lr_intersections(gifti, gifti_relabeled)

# initialize parcellator
parc = parcellate.Parcellater(gifti_relabeled, "fsLR")

# extract T1w/T2w ratio defined by ICA parcellation
t1t2_map = datasets.fetch_annotation(source="hcps1200", desc="myelinmap", space="fsLR", den="32k")
t1t2_lr_split = parc.fit_transform(t1t2_map, "fsLR")  # (178, ): 88L <-> 90R

# average regions that cross both L-R hemispheres
t1t2_lr_avg = np.zeros(106)
t1t2_lr_avg[l_idx] = (t1t2_lr_split[l_idx] + t1t2_lr_split[r_idx]) / 2
zero_idx = np.where(t1t2_lr_avg == 0)[0]
t1t2_lr_avg[zero_idx] = t1t2_lr_split[zero_idx]
```

Thank you for your insights,
Gabriel

### Code of Conduct

- [X] I agree to follow the `neuromaps` Code of Conduct

**Comment by justinehansen** [comment]
Hi Gabriel,

Thanks for reaching out 🙂  I don't have a satisfying answer for you but here's my train of thought:
* as you found out, neuromaps isn't great when it comes to parcellations where some parcels span both hemispheres (we should work on this). What you're doing seems reasonable to me
* one thing you might need to worry about is whether your left-right averaging step is biasing the average, since it weighs both the left and right part of the parcel equally but it's possible the left and right part of the parcel are composed of a very different number of vertices. A weighted average based on the ratio of number of vertices within left and right part of the parcel might be better

That said, there might be easier alternatives. Some background: for parcellations where every parcel is hemisphere-specific (let's say a 200 region parcellation), neuromaps' `Parcellater` wants the left hemisphere indices (0-99) to be different from the right hemisphere indices (100-199). Otherwise (let's say left hemisphere and right hemisphere are both organized in parcellation files that go from 0-99), the parcellater will assume everything of index eg 5 is pointing to the same brain region. This is bad in most cases but actually good in your case - you _want_ that everything of index 5 is pointing to the same parcel, even if the vertices exist on both hemispheres. Therefore, `relabel_gifti()` isn't doing you any favours. 

You can forgo the `relabel_gifti()` step, but then the problem is that your indices run from 0-299 but aren't consecutive (there are 107 values total in that range). So for `Parcellater` to work, you need to cut out the gaps so the values go from 0-106 (0 is background). OR (sorry this is getting very convoluted), you can forgo `Parcellater` entirely and just loop through the values in `np.unique(np.concatenate((gifti[0].agg_data(), gifti[1].agg_data())))`. Something like:

```
from neuromaps.images import load_data, dlabel_to_gifti
from neuromaps.datasets import fetch_annotation

dlabel = "nregions-300_cifti.dlabel.nii"
gifti = dlabel_to_gifti(dlabel)
gifti = np.concatenate((gifti[0].agg_data(), gifti[1].agg_data()))
t1t2 = load_data(fetch_annotation(desc='myelinmap'))

t1t2_parc = np.zeros((len(np.unique(gifti)), ))
for i, idx in enumerate(np.unique(gifti)):
    t1t2_parc[i] = np.mean(t1t2[gifti == idx])
t1t2_parc = np.delete(t1t2_parc, 0)
```
That's where I'm at right now but I'll let you know if I think of something better. 

Last thing: could you tell me a bit more about this parcellation? Why is it ok to remove more than half of the parcels? (Just for my own sake, I don't have much experience with CIFTI honestly).

Best,
Justine

**Comment by griegner** [comment]
Hi Justine, 

Thank you for such a detailed and helpful response! It makes a lot of sense that `Parcellator` expects each label to point to hemisphere-specific vertices, and that in this case, we want each label to instead point to vertices in both hemispheres. 

I think relabeling the (0-299) non-consecutive labels to (0-106) consecutive labels might be the best solution, because it would allow me to use the magic of `Parcellator` and apply it to annotations from multiple coordinate systems. So far, it seems like the code below is working as expected :)

For more context on the parcellation, the CIFTI file includes labels for L-cortex (30k vertices), R-cortex (30k vertices), and sub-cortex (30K voxels). The spatial-ICA was applied in this space, but in my case I'm only looking at cortical surface parcels, so it's okay that `dlabel_to_gifti` removes the subcortical voxels. 

```python
import nibabel as nib
import numpy as np
# neuromaps version 0.0.3+24.gabf5a5c
from neuromaps import datasets, images, parcellate


def relabel_gifti(gifti):
    """Updates GIFTI images so label IDs are consecutive integers.

    Parameters
    ----------
    gifti : (2,)
        Surface label files in GIFTI format (lh.label.gii, rh.label.gii).

    Returns
    -------
    gifti_relabelled: (2,) tuple-of-nib.gifti.GiftiImage
        Relabeled `gifti` images.
    """

    gifti_l, gifti_r = gifti[0].agg_data(), gifti[1].agg_data()

    unique_labels = np.unique(np.concatenate([gifti_l, gifti_r]))
    label_map = {old_label: new_label for new_label, old_label in enumerate(unique_labels)}
    label_map = np.vectorize(label_map.get)

    gifti_relabelled_l, gifti_relabelled_r = label_map(gifti_l), label_map(gifti_r)

    # make new gifti image with updated labels
    gifti_relabelled = tuple()
    for hemi in (gifti_relabelled_l, gifti_relabelled_r):
        darray = nib.gifti.GiftiDataArray(
            hemi, intent="NIFTI_INTENT_LABEL", datatype="NIFTI_TYPE_INT32"
        )
        labeltable = nib.gifti.GiftiLabelTable()
        labeltable.labels = np.unique(hemi)
        img = nib.GiftiImage(darrays=[darray], labeltable=labeltable)
        gifti_relabelled += (img,)

    return gifti_relabelled


# cifti surface+volumetric parcellation file
dlabel = "./data/nregions-300_cifti.dlabel.nii"

# cifti -> surface removes volumetric data: 300 -> 106 parcels
gifti = images.dlabel_to_gifti(dlabel)
gifti_relabeled = relabel_gifti(gifti)


# try parcellation of spaces = {fsLR, fsaverage, civet, MNI152}
parc = parcellate.Parcellater(gifti_relabeled, space="fsLR", resampling_target="parcellation")

fsLR = datasets.fetch_annotation(desc="myelinmap")
fsLR_fit = parc.fit_transform(fsLR, space="fsLR")

fsaverage = datasets.fetch_annotation(desc="genepc1")
fsaverage_fit = parc.fit_transform(fsaverage, space="fsaverage")

civet = datasets.fetch_annotation(desc="scalinghcp")
civet_fit = parc.fit_transform(civet, space="civet")

mni152 = datasets.fetch_annotation(source="aghourian2017", desc="feobv")
mni152_fit = parc.fit_transform(mni152, space="MNI152")

assert fsLR_fit.shape == fsaverage_fit.shape == civet_fit.shape == mni152_fit.shape
```

Thanks again,
Gabriel

**Comment by justinehansen** [comment]
Sweet, your code looks good to me! And thanks for the civet info :slightly_smiling_face:
